src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala

```diff
@@ -149,42 +149,49 @@ object FilterIndexRule extends Rule[LogicalPlan] with Logging {
       filterColumns: Seq[String],
       fsRelation: HadoopFsRelation): Seq[IndexLogEntry] = {
 
-    // map of signature provider to signature for this subplan
-    val signatureMap: mutable.Map[String, String] = mutable.Map()
-
-    def signatureValid(entry: IndexLogEntry): Boolean = {
-      val sourcePlanSignatures = entry.source.plan.properties.fingerprint.properties.signatures
-      assert(sourcePlanSignatures.length == 1)
-      val sourcePlanSignature = sourcePlanSignatures.head
-
-      if (!signatureMap.contains(sourcePlanSignature.provider)) {
-        val signature = LogicalPlanSignatureProvider
-          .create(sourcePlanSignature.provider)
-          .signature(project)
-        signatureMap.put(sourcePlanSignature.provider, signature)
-      }
-
-      signatureMap(sourcePlanSignature.provider).equals(sourcePlanSignature.value)
-    }
+    RulesHelper.getLogicalRelation(project) match {
+      case Some(r) =>
```

Comment:
Could you please explain the reason for this change? To me, this looks like hard-coding signature calculation logic to a signature provider which works only on LogicalRelation. I think here we should keep it relaxed to any "plan" node and not just logical relation.

Reply:
This is not done for saving calculations, but it is needed for correctness.
Assume we have a query plan as:
	QueryPlan (P):  Scan(R: a, b, c) -> Filter(a>5) -> Project(b)
and we have two indexes defined as:	
	Covering Index 1 (IX1): R(a, b)
	Index 2 (IX2): Scan(a, b, c) -> Filter(a>100) -> Project(b, c)	
As the LogicalRelation node is the same for all 3 plans, `FileBasedSignatureProvider` generates the same signature for all three of them.
`IndexSignatureProvider` (introduced by this change) includes node names in signature calculation and as a result it generate the same signature for P and IX2 and a different signature for IX1.

In FilterIndex and JoinIndex rules, a candidate index has to meet below two conditions:
- Index has to be defined on the same LogicalRelation as the one query uses,
- Index has to cover all records (i.e. index plan should not have a Filter, Join, Agg or any other node that removes or modifies original relation's records).
    
As a result, when we look for indexes in FilterIndex and JoinIndex rules we need to: "Extract a single logical relation node from original query plan and compare the signature of that subplan with the signature of covering indexes".
This way the signature comparison will satisfy both conditions mentioned above for these rules and only returns covering indexes that include all records of original relation.
In the above example, this procedure of extracting logical relation from query plan and comparing its signature with index signatures (which are calculated and stored using IndexSignatureProvider) results in only marking IX1 as a candidate index.

Please let me know if above explanation makes sense; we can discuss it further if you prefer.